### ISSUE TITLE
Incorporate XSDKDefaults.cmake into TriBITS

### ISSUE THREAD

**Issue by bartlettroscoe** [TriBITSPub/TriBITS#107]
One of the goals of the IDEAS project is to define configuration standards for packages.  The current specification for those standards are defined [here](https://docs.google.com/document/d/18028D6nsuhIrCvJnX6c07r8m_Np4SH-aGXMX4svMs1w/edit).  Part of that effort is to develop a standard CMake module `XSDKDefaults` that will automatically implement most of the standard behaviors.

This TriBITS Story is to track the work to incorporate the XSDKDefaults.cmake module into TriBITS once it is complete.  This will be tested inside of TriBITS using TribitsExampleProject.

**Comment by bartlettroscoe** [comment]
I forgot to update the documentation.  I need to do that before I close this Isuse.

**Comment by bartlettroscoe** [comment]
I would really love to have someone review this.  

@jwillenbring, given that you are on the IDEAS project and the project requires these xSDK configure options, do you think you can review this?  In particular, it would be good to review the text in the documentation that mentions "XSDK".  See:

    https://tribits.org/doc/TribitsDevelopersGuide.html

Thanks,

-Ross

**Comment by bartlettroscoe** [comment]
Here are my notes from the development process over the weekend for archival purposes (but the commit logs in the XSDKCMakeProj and TriBITS repos also reflect this).

-----

The code for XSDKDefaults.cmake is maintained in the independent GitHub repo:

    https://github.com/xsdk-ideas/XSDKCMakeProj

Way back in December 2014, I had prototyped the handling of C++ compiler and options and other options.  But it needs to be completed by doing the same for C and Fortran compilers and options.

1) Investigate how the XSDKDefaults.cmake and automated testing is set up and how to take advantage of that in adding support for C and Fortran [Done]

2) Refactor the XSDKDefault.cmake module and automated tests to make it easier to add support for handling of C and Fortran [Done]

3) Add support for C and Fortran and add unit tests [Done]

4) Snapshot XSDKDefaults.cmake into TriBITS and add automated tests to make sure it is working as expected.

In order to snapshot XSDKDefault.cmake using the TriBITS Snapshot tool, we should really move XSDKDefault.cmake into its own subdir xsdk/ and then snapshot it that way.  [Done]

Also, I want to make it easy to test TriBITS against the version controlled copy of XSDKCMakeProj so that I can do testing before snapshotting.  To do this, I will use the non-cache var ${PROJECT_NAME}_TRIBITS_XSDK_DIR.  Hopefully no one will define that except for explicit testing.

Now to add XSDKDefaults.cmake to TriBITS:

4.a) Add non-cache var ${PROJECT_NAME}_TRIBITS_XSDK_DIR and if set and the dir exists, include XSDKDefaults.cmake into the right place in TriBITS Core. [Done]

4.b) Add at least one automated test to TriBITS to show that XSDKDefaults.cmake is working as expected. [Done]

4.c) Update TribitsBuildReferenceBody.rst for addition of XSDK options [Done]

5) Snapshot TriBITS into Trilinos do some manual testing (see [Trilinos #76](https://github.com/trilinos/Trilinos/issues/76)).

**Comment by bartlettroscoe** [comment]
I realized that I need to specify and test how compilers are selected under TriBITS when `USE_XSDK_DEFAULTS=TRUE`.  I am worried that if the user sets and env var `CC` and the cache var `MPI_C_COMPILER`, for example, that it will ignore the `MPI_C_COMPILER` option.  Or perhaps that is how it should be in MPI mode when XSDK defaults are enabled?  In any case, I need to characterize how this works and then update the documentation to match.

Let's see what happens without XSDK mode when you set both the env var `CC` and the cache var `MPI_C_COMPILER`:

```
$ env CC=`which mpicc` \
  cmake -DTPL_ENABLE_MPI=ON \
  -DMPI_C_COMPILER=/projects/vera/gcc-4.6.1/toolset/openmpi-1.4.3/bin/mpicc \
  ${TRILINOS_DIR} \
  &> configure.out 
```

This produced:

```
-- USE_XSDK_DEFAULTS='FALSE'
-- Setting CMAKE_BUILD_TYPE=RELEASE since it was not set ...
-- CMAKE_BUILD_TYPE='RELEASE'
-- BUILD_SHARED_LIBS='FALSE'
-- MPI_USE_COMPILER_WRAPPERS='ON'
-- MPI_C_COMPILER='/projects/vera/gcc-4.6.1/toolset/openmpi-1.4.3/bin/mpicc'
-- MPI_C_COMPILER='/projects/vera/gcc-4.6.1/toolset/openmpi-1.4.3/bin/mpicc'
-- Setting CMAKE_C_COMPILER=${MPI_C_COMPILER}
-- CMAKE_C_COMPILER='/projects/vera/gcc-4.6.1/toolset/openmpi-1.4.3/bin/mpicc'
-- MPI_CXX_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicxx'
-- Setting CMAKE_CXX_COMPILER=${MPI_CXX_COMPILER}
-- CMAKE_CXX_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicxx'
-- MPI_Fortran_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpif90'
-- Setting CMAKE_Fortran_COMPILER=${MPI_Fortran_COMPILER}
-- CMAKE_Fortran_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpif90'
-- MPI_EXEC='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpiexec'
```

Look lat that, it ignored the env var `CC` and just went with `MPI_C_COMPILERS`.  But not what happened with the C++ and Fortran compilers, it found the GCC 4.8.3 `mpicxx` and `mpif90` compilers in the default path which are *not* consistent with the GCC 4.6.1 `mpicc` compiler.  That is not consistent with how CMake finds compilers.  That is, it should restrict the search directory to the directory for the C++ and Fortran compilers to where the MPI_C_COMPILER is located.

So what happens if you set both `MPI_C_COMPILER` and `CMAKE_C_COMPILER` to different paths:

```
$ rm -r CMake*
$ env CC=/usr/bin/cc \
  cmake -DTPL_ENABLE_MPI=ON \
  -DCMAKE_C_COMPILER=/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicc \
 -DMPI_C_COMPILER=/projects/vera/gcc-4.6.1/toolset/openmpi-1.4.3/bin/mpicc \
 ../../../Trilinos \
 &> configure.out
```

That produced:

```
-- USE_XSDK_DEFAULTS='FALSE'
-- Setting CMAKE_BUILD_TYPE=RELEASE since it was not set ...
-- CMAKE_BUILD_TYPE='RELEASE'
-- BUILD_SHARED_LIBS='FALSE'
-- MPI_USE_COMPILER_WRAPPERS='ON'
-- Leaving current CMAKE_C_COMPILER=/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicc since it is already set!
-- MPI_CXX_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicxx'
-- Setting CMAKE_CXX_COMPILER=${MPI_CXX_COMPILER}
-- CMAKE_CXX_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicxx'
-- MPI_Fortran_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpif90'
-- Setting CMAKE_Fortran_COMPILER=${MPI_Fortran_COMPILER}
-- CMAKE_Fortran_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpif90'
-- MPI_EXEC='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpiexec'
```

Okay, that is good.  At least it `CMAKE_C_COMPILER` takes precedence over `MPI_C_COMPILER`, as it should.

So what happens with `USE_XSDK_DEFAULTS=TRUE`?

```
$ rm -r CMake*
$ env CC=`which mpicc` \
  cmake -DTPL_ENABLE_MPI=ON -DUSE_XSDK_DEFAULTS=ON \
  -DMPI_C_COMPILER=/projects/vera/gcc-4.6.1/toolset/openmpi-1.4.3/bin/mpicc \
  ${TRILINOS_DIR} \
  &> configure.out 
```

That produced:

```
-- USE_XSDK_DEFAULTS='TRUE'
-- XSDK: Setting CMAKE_C_COMPILER from env var CC='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicc'!
-- XSDK: Setting default CMAKE_BUILD_TYPE=DEBUG
-- CMAKE_BUILD_TYPE='DEBUG'
-- BUILD_SHARED_LIBS='TRUE'
-- MPI_USE_COMPILER_WRAPPERS='ON'
-- Leaving current CMAKE_C_COMPILER=/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicc since it is already set!
-- MPI_CXX_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicxx'
-- Setting CMAKE_CXX_COMPILER=${MPI_CXX_COMPILER}
-- CMAKE_CXX_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpicxx'
-- MPI_Fortran_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpif90'
-- Setting CMAKE_Fortran_COMPILER=${MPI_Fortran_COMPILER}
-- CMAKE_Fortran_COMPILER='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpif90'
-- MPI_EXEC='/projects/vera/gcc-4.8.3/toolset/mpich-3.1.3/bin/mpiexec'
```

So I was right.  When in XSDK mode, TriBITS will use the compiler given in the env vars `CC`, `CXX`, and `FC` and ignore the CMake cache vars `MPI_<LANG>_COMPILER`.  I guess that is appropriate to be consistent with the XSDK for now?

In any case, I think the search for the MPI compilers with TriBITS should follow the same behavior as the search for the raw compilers with raw CMake as mentioned above where it searches for the C++ and Fortran compiler only in the same directory as the C compiler (except it should not find the wrong C compiler and then mess everything up).

So I need to decide how to address this inconsistency and confusion.  Whatever I do will break backward compatibility so I have to be careful.  But I suspect that people are setting all three compilers if they are setting any of them so I guess is that I will not break anything major.

**Comment by bartlettroscoe** [comment]
@jwillenbring reviewed this as part of the Trilinos-specific issue trilinos/trilinos#76.  Therefore, I am going to go ahead and close this.  I will create a new Issue ticket for making the MPI compiler find behavior more consistent.
